File: Frontend_new/src/services/xpCalculator.js
/**
 * XP Calculator Service - Frontend
 * Mirrors backend XP calculation for consistent display
 */

// XP Constants
const XP_RATES = {
  DAILY_LOG_BASE: 20,
  DAILY_LOG_SLEEP: 20,      // >= 7 hours
  DAILY_LOG_WATER: 20,      // >= 6 glasses
  DAILY_LOG_EXERCISE: 20,   // moderate/intense
  DAILY_LOG_NUTRITION: 20,  // good/great
  DAILY_LOG_MOOD: 10,       // >= 3/4

  STREAK_WEEKLY: 100,       // per week of streak
  STREAK_MONTHLY: 500,      // per month of streak

  ARTIFACT: 50,             // per unlocked artifact
  CHAPTER: 25,              // per storybook chapter
};

/**
 * Calculate XP from a single daily log entry
 */
export function calculateLogXP(log) {
  if (!log) return 0;
  
  let xp = 0;
  
  // Sleep XP (7+ hours = 20 XP)
  if (log.sleep >= 7) {
    xp += XP_RATES.DAILY_LOG_SLEEP;
  }
  
  // Hydration XP (6+ glasses = 20 XP) 
  if (log.water >= 6) {
    xp += XP_RATES.DAILY_LOG_WATER;
  }
  
  // Exercise XP (moderate/intense = 20 XP)
  if (typeof log.exercise === 'string') {
    if (['moderate', 'intense'].includes(log.exercise)) {
      xp += XP_RATES.DAILY_LOG_EXERCISE;
    }
  } else if (log.exercise >= 2) { // numeric 0-3 scale
    xp += XP_RATES.DAILY_LOG_EXERCISE;
  }
  
  // Nutrition XP (good/great = 20 XP)
  if (typeof log.nutrition === 'string') {
    if (['good', 'great'].includes(log.nutrition)) {
      xp += XP_RATES.DAILY_LOG_NUTRITION;
    }
  } else if (log.nutrition >= 2) { // numeric 0-3 scale
    xp += XP_RATES.DAILY_LOG_NUTRITION;
  }
  
  // Mood XP (3+ = 10 XP)
  if (log.mood >= 3) {
    xp += XP_RATES.DAILY_LOG_MOOD;
  }
  
  return xp;
}

/**
 * Calculate streak XP from log history
 */
export function calculateStreakXP(logs) {
  if (!logs || logs.length === 0) return 0;
  
  // Sort logs by date (newest first)
  const sortedLogs = [...logs].sort((a, b) => new Date(b.date) - new Date(a.date));
  
  let currentStreak = 1;
  let maxStreak = 1;
  
  for (let i = 1; i < sortedLogs.length; i++) {
    const prevDate = new Date(sortedLogs[i-1].date);
    const currDate = new Date(sortedLogs[i].date);
    const daysDiff = Math.floor((prevDate - currDate) / (1000 * 60 * 60 * 24));
    
    if (daysDiff === 1) {
      currentStreak++;
      maxStreak = Math.max(maxStreak, currentStreak);
    } else {
      currentStreak = 1;
    }
  }
  
  // Calculate streak bonuses
  let streakXP = 0;
  
  // Weekly streak bonuses (100 XP per week)
  if (maxStreak >= 7) {
    streakXP += Math.floor(maxStreak / 7) * XP_RATES.STREAK_WEEKLY;
  }
  
  // Monthly streak bonuses (500 XP per month)
  if (maxStreak >= 30) {
    streakXP += Math.floor(maxStreak / 30) * XP_RATES.STREAK_MONTHLY;
  }
  
  return streakXP;
}

/**
 * Calculate total XP breakdown for display
 */
export function calculateTotalXP(data) {
  const { logs = [], artifacts = [], chapters = [] } = data;
  
  // Daily log XP
  const dailyLogXP = logs.reduce((total, log) => total + calculateLogXP(log), 0);
  
  // Streak XP
  const streakXP = calculateStreakXP(logs);
  
  // Museum XP
  const museumXP = artifacts.filter(a => a.unlocked).length * XP_RATES.ARTIFACT;
  
  // Storybook XP
  const storyXP = chapters.length * XP_RATES.CHAPTER;
  
  const totalXP = dailyLogXP + streakXP + museumXP + storyXP;
  
  return {
    totalXP,
    dailyLogXP,
    streakXP,
    museumXP,
    storyXP,
    breakdown: {
      logs_count: logs.length,
      artifacts_count: artifacts.filter(a => a.unlocked).length,
      chapters_count: chapters.length,
      max_streak: calculateMaxStreak(logs)
    }
  };
}

/**
 * Calculate user level from XP
 */
export function calculateLevel(xp) {
  return Math.min(10, Math.floor(xp / 100) + 1);
}

/**
 * Calculate XP needed for next level
 */
export function calculateNextLevelXP(level) {
  return level * 100;
}

/**
 * Calculate current streak from logs
 */
export function calculateCurrentStreak(logs) {
  if (!logs || logs.length === 0) return 0;
  
  const sortedLogs = [...logs].sort((a, b) => new Date(b.date) - new Date(a.date));
  const today = new Date().toISOString().slice(0, 10);
  
  let streak = 0;
  let currentDate = new Date(today);
  
  for (const log of sortedLogs) {
    const logDate = new Date(log.date);
    const expectedDateStr = currentDate.toISOString().slice(0, 10);
    
    if (log.date === expectedDateStr) {
      streak++;
      currentDate.setDate(currentDate.getDate() - 1);
    } else {
      break;
    }
  }
  
  return streak;
}

/**
 * Calculate maximum streak from logs
 */
function calculateMaxStreak(logs) {
  if (!logs || logs.length === 0) return 0;
  
  const sortedLogs = [...logs].sort((a, b) => new Date(a.date) - new Date(b.date));
  
  let currentStreak = 1;
  let maxStreak = 1;
  
  for (let i = 1; i < sortedLogs.length; i++) {
    const prevDate = new Date(sortedLogs[i-1].date);
    const currDate = new Date(sortedLogs[i].date);
    const daysDiff = Math.floor((currDate - prevDate) / (1000 * 60 * 60 * 24));
    
    if (daysDiff === 1) {
      currentStreak++;
      maxStreak = Math.max(maxStreak, currentStreak);
    } else {
      currentStreak = 1;
    }
  }
  
  return maxStreak;
}

/**
 * Get XP progress for display
 */
export function getXPProgress(xp) {
  const level = calculateLevel(xp);
  const levelStartXP = (level - 1) * 100;
  const nextLevelXP = calculateNextLevelXP(level);
  const progressXP = xp - levelStartXP;
  const progressPercent = (progressXP / 100) * 100; // Each level is 100 XP
  
  return {
    level,
    currentXP: xp,
    progressXP,
    nextLevelXP,
    progressPercent: Math.min(100, progressPercent),
    isMaxLevel: level >= 10
  };
}

export default {
  calculateLogXP,
  calculateStreakXP,
  calculateTotalXP,
  calculateLevel,
  calculateNextLevelXP,
  calculateCurrentStreak,
  getXPProgress
};

File: backend/services/xp_calculator.py
from datetime import date, timedelta
from typing import Dict, List
import json
from sqlalchemy.orm import Session
from core.models import DailyLog, SanctuaryState, MuseumArtifact, StorybookChapter


def calculate_user_xp(db: Session, user_id: int) -> Dict:
    """
    Calculate total XP from all sources for a user.
    Returns breakdown of XP sources.
    """
    # Get all daily logs
    logs = db.query(DailyLog).filter(DailyLog.user_id == user_id).all()
    
    # Calculate daily log XP
    daily_log_xp = 0
    for log in logs:
        daily_log_xp += calculate_log_xp(log_to_dict(log))
    
    # Calculate streak XP
    streak_xp = calculate_streak_xp(logs)
    
    # Calculate museum XP
    artifacts = db.query(MuseumArtifact).filter(
        MuseumArtifact.user_id == user_id,
        MuseumArtifact.unlocked == True
    ).all()
    museum_xp = len(artifacts) * 50  # 50 XP per artifact
    
    # Calculate storybook XP
    chapters = db.query(StorybookChapter).filter(StorybookChapter.user_id == user_id).all()
    story_xp = len(chapters) * 25  # 25 XP per chapter
    
    total_xp = daily_log_xp + streak_xp + museum_xp + story_xp
    
    return {
        "total_xp": total_xp,
        "daily_log_xp": daily_log_xp,
        "streak_xp": streak_xp,
        "museum_xp": museum_xp,
        "story_xp": story_xp,
        "breakdown": {
            "logs_count": len(logs),
            "artifacts_count": len(artifacts),
            "chapters_count": len(chapters)
        }
    }


def calculate_log_xp(log_dict: Dict) -> int:
    """Calculate XP awarded for a single daily log"""
    xp = 0
    if log_dict.get("sleep", 0) >= 7:
        xp += 20
    if log_dict.get("water", 0) >= 6:
        xp += 20
    if log_dict.get("exercise", "none") in ("moderate", "intense"):
        xp += 20
    if log_dict.get("nutrition", "poor") in ("good", "great"):
        xp += 20
    if log_dict.get("mood", 0) >= 3:
        xp += 10
    return xp


def calculate_streak_xp(logs: List[DailyLog]) -> int:
    """Calculate bonus XP from consecutive day streaks"""
    if not logs:
        return 0
    
    # Get all daily entries from JSON structure
    all_entries = []
    for log in logs:
        try:
            entries = json.loads(log.daily_entries) if log.daily_entries else {}
            for date_str, entry_data in entries.items():
                all_entries.append((date_str, entry_data))
        except (json.JSONDecodeError, AttributeError):
            continue
    
    if not all_entries:
        return 0
    
    # Sort by date
    sorted_entries = sorted(all_entries, key=lambda x: x[0])
    
    current_streak = 1
    max_streak = 1
    
    for i in range(1, len(sorted_entries)):
        prev_date_str = sorted_entries[i-1][0]
        curr_date_str = sorted_entries[i][0]
        
        try:
            prev_date = date.fromisoformat(prev_date_str)
            curr_date = date.fromisoformat(curr_date_str)
            
            if curr_date == prev_date + timedelta(days=1):
                current_streak += 1
                max_streak = max(max_streak, current_streak)
            else:
                current_streak = 1
        except ValueError:
            current_streak = 1
    
    # Award bonus XP for streaks
    streak_xp = 0
    if max_streak >= 7:
        streak_xp += (max_streak // 7) * 100  # 100 XP per week
    if max_streak >= 30:
        streak_xp += (max_streak // 30) * 500  # 500 XP per month
    
    return streak_xp


def log_to_dict(log: DailyLog) -> Dict:
    """Convert DailyLog to dict for XP calculation"""
    # Handle JSON structure - get the most recent entry
    try:
        entries = json.loads(log.daily_entries) if log.daily_entries else {}
        if entries:
            # Get the most recent entry
            latest_date = max(entries.keys())
            latest_entry = entries[latest_date]
            return {
                "sleep": latest_entry.get("sleep", 0),
                "water": latest_entry.get("water", 0),
                "exercise": latest_entry.get("exercise", 0),
                "nutrition": latest_entry.get("nutrition", 0),
                "mood": latest_entry.get("mood", 0)
            }
    except (json.JSONDecodeError, AttributeError, ValueError):
        pass
    
    # Return default values if no data
    return {
        "sleep": 0,
        "water": 0,
        "exercise": 0,
        "nutrition": 0,
        "mood": 0
    }


def recalculate_sanctuary_xp(db: Session, user_id: int) -> SanctuaryState:
    """Recalculate and update sanctuary XP from scratch"""
    xp_data = calculate_user_xp(db, user_id)
    
    sanctuary = db.query(SanctuaryState).filter(SanctuaryState.user_id == user_id).first()
    if not sanctuary:
        sanctuary = SanctuaryState(user_id=user_id)
        db.add(sanctuary)
    
    sanctuary.xp = xp_data["total_xp"]
    sanctuary.level = calculate_level(sanctuary.xp)
    sanctuary.next_level_xp = calculate_next_level_xp(sanctuary.level)
    
    db.flush()
    return sanctuary


def calculate_level(xp: int) -> int:
    """Calculate level from total XP"""
    return min(10, (xp // 100) + 1)


def calculate_next_level_xp(level: int) -> int:
    """Calculate XP needed for next level"""
    return level * 100